### PR TITLE
Add CopyToClipBoard to context menu on FileHistory form.

### LINKED ISSUE #4877: "Copy to clipboard" context menu items missing in File History
**Do you want to request a *feature* or report a *bug*?**
A feature.

**What is the current behavior?**
The context menu on commits in the File History window is missing the _Copy to clipboard_ submenu which is present in the Browse window.

**What is the expected behavior?**
It would be useful to have the ability to copy the information related to the commit in the File History window, too. This includes commit hash, message, author, date and names of tags and branches (if any correspond to the selected commit).

**Environment you encounter the issue:**

- GitExtensions version: 2.51
- GIT version: 2.17.0.windows.1
- OS version: Windows 10

**Did this work in previous version of GitExtensions (which)?**
No.

### PR DESCRIPTION
**Pull request by deepak--sharma**
Fixes #4877 

Screenshots before and after (if PR changes UI):
before
before

after
after

What did I do to test the code and ensure quality:
- Ran all the tests. 2 tests are breaking even after undoing my changes. This proves my change is not responsible.
image

Has been tested on (remove any that don't apply):
- GIT 2.10 and above
- Windows 7 and above

### CODE REVIEW

File: GitUI/CommandsDialogs/FormFileHistory.cs [hunk 1]
```diff
@@ -583,5 +604,74 @@ private void detectMoveAndCopyInThisFileToolStripMenuItem_Click(object sender, E
             detectMoveAndCopyInThisFileToolStripMenuItem.Checked = AppSettings.DetectCopyInAllOnBlame;
             UpdateSelectedFileViewers(true);
         }
+
+        private void CopyToClipboardToolStripMenuItem_DropDownOpened(object sender, EventArgs e)
+        {
+            var selectedRevisions = FileChanges.GetSelectedRevisions();
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/FormFileHistory.cs]: Some redundency here... I would have introduced a method taking an `Action` like:

```csharp
        private void DoOnSelectedRevision(Action<GitRevision> ToDo)
        {
            var selectedRevisions = FileChanges.GetSelectedRevisions();
            if (selectedRevisions != null)
            {
                var r = selectedRevisions[0];
                if (r != null)
                {
                    ToDo(r);
                }
            }
        }
```

and used in all your methods

**Comment by pmiossec** [comment]
I think that the menu item should be added before the first separator with the other "actions" (and not at the end).

After the `Save As`?

Do you all agree with that? (cc @RussKie )

File: GitUI/CommandsDialogs/FormFileHistory.cs (hunk 1 above)
Comment (deepak--sharma) [reply GitUI/CommandsDialogs/FormFileHistory.cs]: Sure I’ll make that better. Thanks for your inputs.

**Comment by RussKie** [comment]
i would also invert `if`s to reduce nesting

**Comment by deepak--sharma** [comment]
Ah! I'm missing ReSharper ;) I will close this soon.. thanks

**Comment by jbialobr** [comment]
@deepak--sharma Could you share the failure message/details for the AddAsMostRecentAsync_should_add_new_path_as_top_entry test?

**Comment by deepak--sharma** [comment]
@jbialobr 

Test Name:	AddAsMostRecentAsync_should_add_new_path_as_top_entry
Test FullName:	GitCommandsTests.UserRepositoryHistory.LocalRepositoryManagerTests.AddAsMostRecentAsync_should_add_new_path_as_top_entry
Test Source:	C:\Users\Deepak\Documents\GitHub\gitextensions\UnitTests\GitCommandsTests\UserRepositoryHistory\LocalRepositoryManagerTests.cs : line 34
Test Outcome:	Failed
Test Duration:	0:00:00.222

Result StackTrace:	
at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Numeric.NumericAssertions`1.Be(T expected, String because, Object[] becauseArgs)
   at GitCommandsTests.UserRepositoryHistory.LocalRepositoryManagerTests.<AddAsMostRecentAsync_should_add_new_path_as_top_entry>d__5.MoveNext() in C:\Users\Deepak\Documents\GitHub\gitextensions\UnitTests\GitCommandsTests\UserRepositoryHistory\LocalRepositoryManagerTests.cs:line 47
--- End of stack trace from previous location where exception was thrown ---
   at NUnit.Framework.Internal.AsyncInvocationRegion.AsyncTaskInvocationRegion.WaitForPendingOperationsToComplete(Object invocationResult) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\AsyncInvocationRegion.cs:line 113
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunAsyncTestMethod(TestExecutionContext context) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 96
Result Message:	Expected newHistory.Count to be 5, but found 4.

**Comment by jbialobr** [comment]
Thanks!

**Comment by RussKie** [comment]
@deepak--sharma I think you may need to debug this one, I can't see anything obvious that can break this test.

**Comment by jbialobr** [comment]
@RussKie I think it is due to AppSettings.RecentRepositoriesHistorySize. If he has it set to 4, the historySize will never be 5. You set it 30 but in the TearDown method.

**Comment by RussKie** [comment]
Aye, this may explain it, thanks for the hint.
There may be few ways of addressing it. Ideally there shouldn't be a coupling between `AppSettings` and `LocalRepositoryManager`. 
In a long run, `AppSettings` should be injectable, as a tactical fix the actual value may somehow need to be passed as a parameter...

**Comment by deepak--sharma** [comment]
@pmiossec , @RussKie all review comments done. Can you please merge? If all good. Thanks.

**Comment by pmiossec** [comment]
@deepak--sharma There are some things not good. I made a PR that you should merge to get the fixes

File: GitUI/CommandsDialogs/FormFileHistory.cs [hunk 2]
```diff
@@ -583,5 +604,67 @@ private void detectMoveAndCopyInThisFileToolStripMenuItem_Click(object sender, E
...
+
+            if (selectedRevisions != null)
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/FormFileHistory.cs]: You must use `DoOnSelectedRevision()`also here....

File: GitUI/CommandsDialogs/FormFileHistory.cs [hunk 3]
```diff
@@ -583,5 +604,67 @@ private void detectMoveAndCopyInThisFileToolStripMenuItem_Click(object sender, E
...
+                    GitUI.UserControls.RevisionGridClasses.CopyToClipboardMenuHelper.AddOrUpdateTextPostfix(authorCopyToolStripMenuItem, r.Author);
+                    GitUI.UserControls.RevisionGridClasses.CopyToClipboardMenuHelper.AddOrUpdateTextPostfix(dateCopyToolStripMenuItem, r.CommitDate.ToString());
+                }
+            }
+        }
+
+        private void HashCopyToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            var selectedRevisions = FileChanges.GetSelectedRevisions();
+            if (selectedRevisions != null)
```
Comment (pmiossec) [inline GitUI/CommandsDialogs/FormFileHistory.cs]: and here...

(Merge my PR to get it fixed!)

**Comment by pmiossec** [comment]
That's not introduced by this PR but I wonder why we copy the "Author" field (and not the commiter) but the "Commit date" (and not the "Author date", which is the one displayed in the grid).

It makes no sense for me. I think we should copy the "author date" or perhaps the 2 when they are different...

What do you think @RussKie and @jbialobr ?

File: GitUI/CommandsDialogs/FormFileHistory.cs [hunk 4]
```diff
@@ -468,6 +469,31 @@ private void FileHistoryContextMenuOpening(object sender, CancelEventArgs e)
             manipulateCommitToolStripMenuItem.Enabled =
                 selectedRevisions.Count == 1 && !selectedRevisions[0].IsArtificial;
             saveAsToolStripMenuItem.Enabled = selectedRevisions.Count == 1;
+
+            // clipboard branch and tag menu handling
+            {
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormFileHistory.cs]: please remove unnecessary braces
move the new implementation into own method

File: GitUI/CommandsDialogs/FormFileHistory.cs [hunk 5]
```diff
@@ -468,6 +469,31 @@ private void FileHistoryContextMenuOpening(object sender, CancelEventArgs e)
...
+                var revision = selectedRevisions[0];
+                copyToClipboardToolStripMenuItem.Enabled = !revision.IsArtificial;
+
+                if (copyToClipboardToolStripMenuItem.Enabled)
+                {
+                    var gitRefListsForRevision = new UserControls.RevisionGridClasses.GitRefListsForRevision(revision);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormFileHistory.cs]: This introduces a coupling to `RevisionGrid` specific implementations, which is unacceptable. 
Neither implementation, which this dependency is referenced for, is complex to justify the coupling. Please inline "get branches" and "get tags"

**Comment by RussKie** [comment]
> It makes no sense for me. I think we should copy the "author date" or perhaps the 2 when they are different... 

We should do the same as we do in the revision graph

**Comment by RussKie** [comment]
@deepak--sharma I pushed a fix for `AddAsMostRecentAsync_should_add_new_path_as_top_entry` tests.
Please check #4921 to see if it fixes the failure you observed.

**Comment by pmiossec** [comment]
>>It makes no sense for me. I think we should copy the "author date" or perhaps the 2 when they are different...

>We should do the same as we do in the revision graph

@Russkie Indeed and if it should be changed, it won't be in this PR.

I would just like your point of view on the subject because for me it make no sense.

We display in the grid the information that we think is the more important for a commit and when we propose the feature to copy it in the paperclip, we copy the other value. Where is the logic?

And that's very strange because we end up with a value that we don't know where it came from.

I think it's even more disturbing for most of the user that don't now the difference between author and commiter notions...

**Comment by RussKie** [comment]
@pmiossec would you mind raising a new issue for this?
This is one these rare cases where I don't a strong opinion/preference because I don't use any of these context menus. As long as the behaviour is consistent and makes sense to the users who use these - I am happy.

**Comment by szaliszali** [comment]
I feel a little sad about this PR because it duplicates (copypastes) a lot of code from `RevisionGrid` reducing the project's health

I think the Copy to clipboard context menu should have been extracted as a reusable component first and then added to this form

**Comment by drewnoakes** [comment]
@deepak--sharma now that #4925 is merged do you feel there's anything in this PR you want to move ahead with, or can it be closed?

**Comment by deepak--sharma** [comment]
@drewnoakes yeah it can be closed now.

**Comment by drewnoakes** [comment]
Thanks Deepak.

**Pull request closed without merging**